core/drivers/imx_sc_api.c

```diff
@@ -0,0 +1,438 @@
+
+/* RNG SECO states */
+enum sc_seco_rng_status {
+	SC_SECO_RNG_STAT_UNAVAILABLE = 0,
+	SC_SECO_RNG_STAT_INPROGRESS,
+	SC_SECO_RNG_STAT_READY
+};
+
+/* Resources IDs */
+enum sc_resource {
+	SC_RES_CAAM_JR1 = 500,
+	SC_RES_CAAM_JR2,
+	SC_RES_CAAM_JR3,
+	SC_RES_CAAM_JR1_OUT = 514,
+	SC_RES_CAAM_JR2_OUT,
+	SC_RES_CAAM_JR3_OUT,
+	SC_RES_CAAM_JR0 = 519,
+	SC_RES_CAAM_JR0_OUT,
+	SC_RES_LAST = 546
+};
+
+/* Power modes */
+enum sc_power_mode {
+	SC_PM_PW_MODE_OFF = 0,
+	SC_PM_PW_MODE_STBY,
+	SC_PM_PW_MODE_LP,
+	SC_PM_PW_MODE_ON
+};
+
+/* RPC message header */
+struct sc_rpc_msg_header {
+	uint8_t version;	/* SC RPC version */
+	uint8_t size;		/* Size of the message */
+	uint8_t svc;		/* Type of service (RM, PM, ...)*/
+	uint8_t func;		/* Function ID or Error code */
+};
+
+/* RPC message format */
+struct sc_rpc_msg {
+	struct sc_rpc_msg_header header;
+	union {
+		uint32_t u32[SC_RPC_MAX_MSG - 1];
+		uint16_t u16[(SC_RPC_MAX_MSG - 1) * 2];
+		uint8_t u8[(SC_RPC_MAX_MSG - 1) * 4];
+	} data;
+};
+
+static struct mutex scu_mu_mutex = MUTEX_INITIALIZER;
+static vaddr_t secure_ipc_addr;
+
+register_phys_mem(MEM_AREA_IO_SEC, SC_IPC_BASE_SECURE, SC_IPC_SIZE);
+
+/*
+ * Read a message from an IPC channel
+ *
+ * @ipc	IPC channel
+ * @msg	Received message
+ */
+static TEE_Result sc_ipc_read(vaddr_t ipc, struct sc_rpc_msg *msg)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	unsigned int count = 0;
+
+	if (!msg) {
+		EMSG("msg is NULL");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	res = mu_receive_msg(ipc, 0, (uint32_t *)msg);
+	if (res)
+		return res;
+
+	/* Check the size of the message to receive */
+	if (msg->header.size > SC_RPC_MAX_MSG) {
+		EMSG("Size of the message is > than SC_RPC_MAX_MSG");
+		return TEE_ERROR_BAD_FORMAT;
+	}
+
+	for (count = 1; count < msg->header.size; count++) {
+		res = mu_receive_msg(ipc, count % MU_NB_RR,
+				     &msg->data.u32[count - 1]);
+		if (res)
+			return res;
+	}
+
+	return TEE_SUCCESS;
+}
+
+/*
+ * Write a message to an IPC channel
+ *
+ * @ipc	IPC channel
+ * @msg	Send message pointer
+ */
+static TEE_Result sc_ipc_write(vaddr_t ipc, struct sc_rpc_msg *msg)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	unsigned int count = 0;
+
+	if (!msg) {
+		EMSG("msg is NULL");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	if (msg->header.size > SC_RPC_MAX_MSG) {
+		EMSG("msg->size is > than SC_RPC_MAX_MSG");
+		return TEE_ERROR_BAD_FORMAT;
+	}
+
+	res = mu_send_msg(ipc, 0, *(uint32_t *)msg);
+	if (res)
+		return res;
+
+	for (count = 1; count < msg->header.size; count++) {
+		res = mu_send_msg(ipc, count % MU_NB_TR,
+				  msg->data.u32[count - 1]);
+		if (res)
+			return res;
+	}
+
+	return TEE_SUCCESS;
+}
+
+/*
+ * Send an RPC message over the secure world IPC channel
+ *
+ * @msg		Message to send. This pointer will also return the answer
+ *		message if expected.
+ * @wait_resp	Set to true if an answer is expected.
+ */
+static TEE_Result sc_call_rpc(struct sc_rpc_msg *msg, bool wait_resp)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+
```

Comment:
`assert(secure_ipc_addr);` to ensure `imx_sc_driver_init()` was previously successfully called?

Reply:
I would check ```vaddr_t ipc``` input parameter of ```sc_ipc_write/read()``` instead